cpp/src/groupby/hash/groupby.cu

```diff
@@ -68,8 +81,248 @@ bool constexpr is_hash_aggregation(aggregation::Kind t) {
+  experimental::detail::initialize_with_identity(table_view, aggs, stream);
+
+  // prepare to launch kernel to do the actual aggregation
+  auto d_sparse_table = mutable_table_device_view::create(sparse_table);
+  auto d_values = table_device_view::create(flattened_values);
+  rmm::device_vector<aggregation::Kind> d_aggs(aggs);
+
+  bool skip_key_rows_with_nulls = keys_have_nulls and ignore_null_keys;
+
+  experimental::detail::grid_1d grid(keys.num_rows(), 256);
+  if (skip_key_rows_with_nulls) {
+    auto row_bitmask{bitmask_and(keys, rmm::mr::get_default_resource(), stream)};
+    hash::compute_single_pass_aggs<true>
+      <<<grid.num_blocks, grid.num_threads_per_block, 0, stream>>>(
+        map, keys.num_rows(), *d_values, *d_sparse_table, d_aggs.data().get(),
+        static_cast<bitmask_type*>(row_bitmask.data()));
+  } else {
+    hash::compute_single_pass_aggs<false>
+      <<<grid.num_blocks, grid.num_threads_per_block, 0, stream>>>(
+        map, keys.num_rows(), *d_values, *d_sparse_table, d_aggs.data().get(),
+        nullptr);
+  }
+
+  // Add results back to sparse_results cache
+  auto sparse_result_cols = sparse_table.release();
+  for (size_t i = 0; i < aggs.size(); i++) {
+    sparse_results.add_result(col_ids[i],
+                              std::make_unique<aggregation>(aggs[i]),
+                              std::move(sparse_result_cols[i]));
+  }
+}
+
+template <bool keys_have_nulls>
+auto groupby_null_templated(
+    table_view const& keys, std::vector<aggregation_request> const& requests,
+    experimental::detail::result_cache& cache,
+    bool ignore_null_keys, cudaStream_t stream,
+    rmm::mr::device_memory_resource* mr)
+{
+  auto d_keys = table_device_view::create(keys);
+  auto map = create_hash_map<keys_have_nulls>(*d_keys, ignore_null_keys, stream);
+
+  // Cache of sparse results where the location of aggregate value in each
+  // column is indexed by the hash map
+  experimental::detail::result_cache sparse_results(requests.size());
+
+  // Compute all single pass aggs first
+  compute_single_pass_aggs<keys_have_nulls>(
+    keys, requests, sparse_results, *map, ignore_null_keys, stream);
+
+  // Now continue with remaining multi-pass aggs
+  // <placeholder>
+
+  // Extract the populated indices from the hash map and create a gather map.
+  // Gathering using this map from sparse results will give dense results.
+  rmm::device_vector<size_type> gather_map(keys.num_rows());
+  rmm::device_scalar<size_type> num_groups(0);
+  experimental::detail::grid_1d grid(keys.num_rows(), 256);
+  extract_gather_map<<<grid.num_blocks, grid.num_threads_per_block, 0, stream>>>
+    (*map, gather_map.data().get(), num_groups.data());
+  gather_map.resize(num_groups.value());
+
+  // Compact all results from sparse_results and insert into cache
+  sparse_to_dense_results(requests, sparse_results, cache, gather_map, stream, mr);
+
+  // Extract unique keys and return
+  auto unique_keys = experimental::detail::gather(
```

Comment:
you can return the gather result directly.

Reply:
This was just to improve code readability. Makes it clear that the result of the gather is unique keys and that's what we return.

Reply:
Hm, then the comment in the line above is redundant.